### ISSUE TITLE
Reconsider restrictions on Atomics.wait in the browser thread

### ISSUE THREAD

**Issue by OleksandrChekhovskyi** [tc39/ecmascript_sharedmem#100]
Consider an application whose main loop runs in a pthread-worker (fully synchronous, blocking, never returns control to a browser), and event handlers on the browser thread post all events to the app main thread through shared memory. This would be a natural model for a direct port of a native application.

The problem comes from actions that are affected by the security mechanism, and are only allowed to execute within event handlers (such as requestFullscreen). A workable solution for this would be to do a blocking wait in the event handler, waiting for the app main thread to process it, while processing any requests coming back, so if app main thread decides that event should trigger requestFullscreen, it could post back such call, and it will be executed within the event handler, as required for security.

This pattern is, obviously, best implementable with Atomics.wait, or otherwise it'll have to spin, which seems wasteful. Alternatively, intrusive changes to the application will be needed to extract the special logic for certain events from the app main thread to the browser thread. For example, if there is a button rendered by the native code that triggers fullscreen (think, in a game), it would need to tell the browser thread where exactly that button is, in advance, for the event to work. That is very clunky.

Also a general argument - if JS can already busy-loop, and browser is expected to know how to interrupt that, what's the problem with Atomics.wait, that can be more energy-friendly?

**Comment by taisel** [comment]
I'd let someone else fully explain the rationalization below, I'll just give my perspective on it. It's a bit over-encumbered on the main thread, and interactivity is paramount with resistance to an additional form of stalling, such as deadlock (Don't need additional tuning to decide when to pop up the stuck script dialog).

That being said, I'm curious to why there is no rAF within webworkers, specifically ones that can be used as a timer per se (to buffer to shm), rather than for drawing. One can already jimmy up a wake from main and a wait in a worker to proxy such, but that's ugly and adds latency.

**Comment by lars-t-hansen** [comment]
@OleksandrChekhovskyi, many will not dispute that there are interesting use cases for Atomics.wait on the main thread, but there is virtually a zero probability that Atomics.wait will be allowed on the main thread in the first version of this technology - blocking the main thread is much too controversial.  The fact that you can busy-wait on the main thread to have some of the same effect does not remove the controversy.

The pattern that I see evolving is one where the master (the main thread) can communicate with the workers by Atomics.wake(), if it needs to, but the workers communicate with the master by postMessage().  I think that eventually async/await support in browsers will make this pattern tolerably easy to use in the master, though perhaps not appropriate for very high message volumes or very low latencies (microseconds).  As far as the performance of that solution is concerned, Issue #64 is probably relevant, especially in connection with Issue #87.

@taisel, I'm afraid your question about requestAnimationFrame makes no sense to me.

(Closing this now because the issue really is closed, but I'll followup to further messages here.)

**Comment by OleksandrChekhovskyi** [comment]
I've tried to look for discussions on this issue, and found this so far:
https://github.com/tc39/ecmascript_sharedmem/issues/36
(is there more discussions elsewhere that I could read?)

It seems like the major issue here is that the user application can break if it deadlocks (note, not the browser, but the app). It's the broken app that will receive the blame from the user in this case, and not the browser.

To me that means we could still allow it on browser thread while explaining to developers that they have to be careful in order to not deadlock in certain scenarios. My position is that we give power with responsibility, instead of taking it away because people might shoot themselves in the leg. The whole shared memory concept is full of shoot-yourself-in-the-leg things already, but it gives developers new useful powers, which is why we want it in the first place.

The use case that I have described (with certain actions that **need** to happen in the event handler), does not seem to be solvable in any other good way. So it will have to wait, but instead of using efficient Atomics.wait, it'll have to do a silly busy loop. Seems impractical.

**Comment by lars-t-hansen** [comment]
Issue #36 has the main points of the older discussion, yes.

You are wrong about the causes of a lockup, because that's the fault of the browser (though the user program may get the blame anyhow).

The major issue is that the user application can break because of browser implementation decisions.  The main thread in a browser performs work on behalf of other threads, notably, on behalf of worker threads.  Consider network I/O as an example, and assume that the main thread performs all I/O on behalf of all threads (this is plausible if not necessarily true in all browsers).  If the main thread is blocked in the user program through Atomics.wait, but the worker thread must do some I/O before it can unblock the blocked main thread with Atomics.wake, then the implementation of I/O in the worker will subcontract that work to the main thread.  But the main thread is blocked and can't perform the work until it's woken.  So the user program deadlocks even though it is correct.  This can probably be fixed, but fixing it requires broad restructuring of the browser, and it's not going to happen.

The major issue for many other participants in the discussion is that the run-to-completion model on the main thread is considered a very important good, and is not something they want to give up.  You may see this as just a technical restriction but others think this model has been fundamental to making asynchronous programming on the web successful (and is IMO why you're seeing async/await in future ECMAScript versions and not eg classical coroutines).

Again, the model that is going to come out of this work is going to be a model where there is a cluster of closely coupled workers using Atomics.wait/wake for signaling and shared memory for data transfer, along with a more loosely coupled main thread that sends signals to the workers with Atomics.wake and receives them through messages, but which can also use shared memory for data transfer.  For some tasks it will be a challenge to make that work well, and maybe we'll need to add some refinements to it.  But blocking on the main thread is not going to be one of the solutions.

**Comment by OleksandrChekhovskyi** [comment]
For the sake of completeness, what if we require a correct user application using Atomics.wait on main thread to always have a timeout there, regardless of what's happening in other threads?

That would solve the event handlers case - they would still wait, but only with a small timeout, and so will never cause deadlocks. In the case of misfortune where application thread was blocking on proxied IO, that event would simply timeout and fail to invoke the desired action (and only if that action was affected by security restrictions).

**Comment by lars-t-hansen** [comment]
Right, so I considered the possibility of an enforced timeout, though I don't think it was ever seriously discussed, and I don't think it works well.  Consider the previous scenario: the worker needs to do I/O but the main thread is blocked.  The worker must wait for the main thread to become unblocked.  The required timeout happens in the main thread and Atomics.wait returns - and then what?  The main thread has three options.  One, it can figure there's an error and throw an exception (shouldn't have timed out, browser's fault, page is not working, try again).  Two, it can retry the wait - after all, what's stopping it?  Now the worker is still not getting I/O done.  Three, it can back out to the event loop to allow the I/O to take place and then ... not clear what it would do.  After all, if it had a strategy for this eventuality it would not need to Atomics.wait in the first place, it could just back out to the event loop and wait for a message or something of that kind.

The reality is that if the main thread cannot block indefinitely then it should not block at all (except maybe for the corner case in #64), because it will need some type of mechanism to continue to function when its timeout expires, which we should expect to happen occasionally.  The alternative is that pages start to fail seemingly at random, as the timeout is reached seemingly at random, under heavy load or simply when something was paged out and needs to be paged in again.  (Data are paged out, the disk spins down - and when you try to get things going again a second or two can easily pass.)

I know the emscripten team has been working on something where proxying code is generated for you automatically, but I don't know how well developed that is.  Let's ask: @kripken, @juj.

**Comment by OleksandrChekhovskyi** [comment]
What I'm advocating for is not "throw an exception if timeout is missing", but instead, dropping all restrictions, and instead writing in the spec "It is a responsibility of a developer to use Atomics.wait on the browser thread in such a way that will not deadlock, even in case if some browser features require main browser event loop in order to function. A program that fails to do that might get aborted by the browser.". And then the browser implementor's responsibility is to ensure that misbehaving application is stopped if deadlock occurs.

This approach changes definition of what is a "correct program" to account for real world browser implementations (in a perfect world, browsers would just provide something like `window.processEvents(timeout)`, and then everyone would live happily and synchronously ever after).

> The reality is that if the main thread cannot block indefinitely then it should not block at all

This reality is exactly what I'm challenging here, with the use case of event handlers. These have a perfectly clear defined behavior in case if application thread does not respond in time - event will simply be dropped, and user will have to try again. In the real world interactive applications, this would be actually extremely rare (doing IO in the main thread has always been discouraged).

And it's not a fringe use case either. This literally popped up immediately when I started porting my game with the question - "how do I get my fullscreen button to work?". Another (very common) thing is to open link in a new tab or a popup window, which also has to happen within an event handler. Probably clipboard access also needs this (haven't checked but it would make sense). Literally any UI application running in a dedicated worker/thread will hit this case pretty much immediately.

And right now we are condemning this use case to do a busy-wait.

**Comment by taisel** [comment]
@lars-t-hansen

Lack of proper vsync eventing in webworker still.

**Comment by lars-t-hansen** [comment]
@OleksandrChekhovskyi,

>  and instead writing in the spec "It is a responsibility of a developer to use Atomics.wait on the browser thread in such a way that will not deadlock, even in case if some browser features require main browser event loop in order to function. A program that fails to do that might get aborted by the browser."

I don't think that type of prose has any chance of making it into the spec, sorry.  It creates a tremendous amount of uncertainty for the web developer and makes compatibility across browsers very hard, for one thing.

In practice, applications that do parallel computation will end up being partitioned between a "front end" written in JS running on the main thread and handling the UI, and a "back end" doing the computation.  These are going to have to be loosely coupled as previously outlined.  If a UI action or event depends on synchronous response from the computation then that's going to be a problem, but that's a limitation I don't think will be lifted.

I suppose this is similar to doing server-side processing (except with much lower latency): a UI action on the client leads to a state where a message is sent and the response will come back later and will then have to be reflected in a UI update.  Some types of programs are not suited to that; others can be adapted to that style.

**Comment by OleksandrChekhovskyi** [comment]
Ok, that was indeed an ill-defined prose. But maybe we can do better? For example,
"in agent with AgentCanSuspend == false, correct user program must guarantee to eventually stop waiting regardless of what happens in other agents". Or something like this.

The thing is, it's already a de-facto requirement, since developer can still choose to wait with a busy-loop, and still would need to take care not to dead-lock in such case. Maybe it's better to explicitly explain it somehow in the spec?

Also, if accidental abuse is a concern, how about something like `Atomics.iKnowWhatIAmDoingAndPromiseNotToDeadlockLetMeWait()`?

It's not like I want to do sync things in the browser thread just to screw with the system, it's because I have to, for things to work correctly. The cosmic irony - to make an interactive UI application work, it has to synchronously block the main thread.

**Comment by lars-t-hansen** [comment]
The user program already has to guarantee to stop eventually (on the main thread) - otherwise the browser will just kill the script, as you also observe.  So I don't know what that prose accomplishes.

The thing this will always come back to is that if you can't guarantee that it's ok to wait indefinitely, then the program needs to handle not being able to wait any longer than it's waited already; and if it has to handle that, then it might as well not be allowed to wait at all.  Building script-aborting into the computational model, which is what you are suggesting as an alternative, is something nobody else wants to do.

You haven't said specifically what the semantics of  `Atomics.iKnowWhatIAmDoingAndPromiseNotToDeadlockLetMeWait()` would be, but to me it looks like it is just an assignment statement that sets AgentCanSuspend to true (whether for good or in a dynamic region of the program).  That would clearly enable deliberate abuse - I can make your browser lock up by having you visit my web site.  Remember, your browser is not allowed to kill my script - my script has promised not to deadlock.  Of course, my script is lying, but how can you tell?

**Comment by OleksandrChekhovskyi** [comment]
Browser should never trust the script, obviously, so even if script promised to use `Atomics.wait` responsibly, but deadlocks, browser should still terminate it. The idea was to protect programmer from accidentally using `Atomics.wait` on main thread without understanding the exact implications, by making it opt-in on behalf of the app. That is, if protecting programmer was the main concern of putting in these restrictions.

I think I have exhausted the arguments by this point - the whole case has been presented in full.
It all boils down to this: it'll need to wait no matter what, might as well let it `Atomics.wait` to not waste CPU cycles in a busy-loop.

**Comment by juj** [comment]
Hey @OleksandrChekhovskyi, I noticed you are working with Emscripten as well, and are probably approaching the question from that perspective. Even though `Atomics.wait` is not allowed on the main thread, we do emulate this in Emscripten, and the function `emscripten_futex_wait` is allowed on the main thread. This basically gives you the function `Atomics.iKnowWhatIAmDoingAndPromiseNotToDeadlockLetMeWait()`.

The above support is implemented via emulation by busy-spinning the main thread. The only drawback of this compared to actually having `Atomics.wait` on the main thread is that the busy-spinning consumes  more CPU than actual wait, but given that main thread only ever expects to sleep for tiny time slices at a time, that is not much of a concern.

Of course all of this is a bit silly, and it will just switch deadlocks in buggy code from one place to other, but if it appeases people evaluating the spec, then that's all good. Well-behaving Emscripten citizens would never call `emscripten_futex_wait`, but it is there and works if you need it.

The real big problem is what Lars mentions in "The major issue is that the user application can break because of browser implementation decisions. The main thread in a browser performs work on behalf of other threads, notably, on behalf of worker threads.". This problem exists whether we busy wait on the main thread or not. The reason why this is such a big problem is that traditionally no JavaScript APIs have been specced from the perspective of how they are allowed to interact internally with other browser threads, since that concept is not visible in the specifications. Two concrete examples of this effect in action are

https://bugzilla.mozilla.org/show_bug.cgi?id=1049091 - Calling console.log/.error in a web worker blocks until the main JS thread yields back to the browser.

https://bugzilla.mozilla.org/show_bug.cgi?id=1049079 - Main JS thread needs to yield back to browser for an indeterminate time in order for a web worker to launch.

In the first bug, calling console.log() in a worker would halt the worker until the main thread has finished executing user JS code. In the second but, calling 'new Worker()' in a worker would halt the worker until the main thread has loaded up the worker and it is running.

The above bugs are something that were not observable in JavaScript at all before shared memory was introduced. In an ideal world, it would be nice to have all JavaScript specs contain a wording that all functions in all APIs shall be able to complete independently without the assistance of the main browser thread (i.e. "truly" multithreaded), but because we have a very long history of existing JavaScript API development, it would be an extremely tall order to require each and every API to be audited to cope with this new multithreading requirement.

(As a sidenote, we also have the same problem in mini-scope inside Emscripten runtime itself. Since for example our filesystem is not yet really multithreaded, all file i/o operations from pthreads are proxied to be completed on the main thread. This generates the same kind of deadlock opportunity, if the main thread is busy-spinning to wait for a message from the pthread, but that pthread has halted to wait for the main thread to perform a proxied operation for this. This issue is formalized in https://github.com/kripken/emscripten/issues/3495#issuecomment-140211370)

The ultimate solution we are slowly working towards is that we'll free up the main browser thread altogether, and move Emscripten main thread to run in a worker. That is, we'll make the --proxy-to-worker mode the default for multithreaded builds. In that mode, the main browser thread will not be running asm.js code, but is always idle, and free to assist the workers in whatever API calls they do. The asm.js main thread will run in a worker. This will immediately resolve bug 1049079. The intent is to proxy all those operations, e.g. full screen requests, over to the main thread, on the level of the C/C++ APIs, i.e. Emscripten HTML5 API (system/include/emscripten/html5.h), SDL, GL, GLEW, and so on. This proxying will be synchronous, and there will be some amount of latency, but everything will work correctly.

The remaining question then is how much latency exactly we are talking about there, and what to do about that, but it is something we should implement and measure to have more informed direction to go with. If some operations/API calls have too much latency, we can start arguing that they should be added to Web Workers directly, so that they don't need to be proxied. That kind of JS spec writing work will be easier to get approved across browser vendors compared to the above "make all JS APIs truly multithreaded" modification.

**Comment by OleksandrChekhovskyi** [comment]
@juj

> Even though Atomics.wait is not allowed on the main thread, we do emulate this in Emscripten, and the function emscripten_futex_wait is allowed on the main thread.

Indeed, I am well aware of this, and this is exactly what I'm making use of already, to implement the case I described.

This line [right here](https://github.com/kripken/emscripten/blob/2c36eebd9faf5a584bd8df9d27a5078500aacd1a/src/library_pthread.js#L794) is literally the whole story of this issue. `Atomics.wait` is not allowed, _but we're going to do it anyway_.

> but given that main thread only ever expects to sleep for tiny time slices at a time, that is not much of a concern.

For the event handlers case, it could sleep for up to one frame interval (say, 1s/60 ~= 16 ms, or possibly even more if framerate is lower). This will only kick in for mouse up/down and key up/down events (as only those can trigger security-affected actions), but still can waste a lot of CPU cycles. One frame interval is a long time from CPU perspective.

> Of course all of this is a bit silly, and it will just switch deadlocks in buggy code from one place to other, but if it appeases people evaluating the spec, then that's all good.

That's basically how I feel about all this. We're adding these restrictions not because there is a good technical case for them, but because somewhere in the _Holy (Java-)Scripture_ it says "thou shalt not sync". From my practically-oriented perspective, this is top-grade dogmatic silliness.

> That is, we'll make the --proxy-to-worker mode the default for multithreaded builds. In that mode, the main browser thread will not be running asm.js code, but is always idle, and free to assist the workers in whatever API calls they do.

I don't think that's even necessary - everything is already possible. In fact, I even prefer to be in control of all proxying myself. In the prototype port of my tech, browser thread sets up event handlers, starts a pthread to host the real app main thread (which is synchronous), and bails with `emscripten_exit_with_live_runtime()`. The idea is that I use current set of emscripten APIs in pure async mode on the browser thread (so normally it's fully available to do anything it needs to do), and application can post callbacks to run on the browser thread to use things that require browser event loop and/or DOM access (btw, for that I have a pull request open, @kripken assigned it to you: https://github.com/kripken/emscripten/pull/4325, and there is also another one, for transferring OffscreenCanvas, https://github.com/kripken/emscripten/pull/4303).

For my comfortable existence as a developer, I only need primitives, and I could construct anything I need. I think that proxying should be managed by the application, and not baked in Emscripten (optional, perhaps?).
